crates/wasm-encoder/src/custom.rs

```diff
@@ -30,6 +23,45 @@ impl Section for CustomSection<'_> {
     }
 }
 
+impl Section<ModuleSectionId> for CustomSection<'_> {
+    fn id(&self) -> ModuleSectionId {
+        ModuleSectionId::Custom
+    }
+
+    fn encode<S>(&self, sink: &mut S)
+    where
+        S: Extend<u8>,
+    {
+        self.encode(sink);
+    }
+}
+
+impl Section<AdapterModuleSectionId> for CustomSection<'_> {
+    fn id(&self) -> AdapterModuleSectionId {
+        AdapterModuleSectionId::Custom
+    }
+
+    fn encode<S>(&self, sink: &mut S)
+    where
+        S: Extend<u8>,
+    {
+        self.encode(sink);
+    }
+}
+
+impl Section<ComponentSectionId> for CustomSection<'_> {
+    fn id(&self) -> ComponentSectionId {
+        ComponentSectionId::Custom
+    }
+
+    fn encode<S>(&self, sink: &mut S)
+    where
+        S: Extend<u8>,
+    {
+        self.encode(sink);
+    }
+}
```

Comment:
I've got two questions from this:

* Why is there a distinction between component sections and adapter module sections? I thought we only had adapter modules? I may be tripping myself up again in perceptions around what I thought this was supposed to be...
* The intention, I thought, was that the component model would share the binary encoding with core wasm wherever possible, so I would hope that we don't need to parameterize the `Section` trait and have three implementations for `CustomSection`, instead just one?

Reply:
Right now the module linking proposal defines _adapter modules_ that are limited to describing how core/adapter modules link together; they only have a small subset of the sections from a core module (type, import, and export) and thus can't really define their own items other than modules and instances (or alias core items from instances).

The component model proposal extends the module linking proposal by overlaying support for interface types; it is a true extension, but with a different version in the header, so they could technically be combined and indeed this is something that Luke was contemplating doing as the utility of having the less capable _adapter module_ format is questionable (perhaps useful for engines that don't want to support the component model at all but want a mechanism to link core modules together?).

The current intention of the binary encoding of both proposals is to make a break from the core binary encoding. Perhaps [this explainer](https://github.com/WebAssembly/interface-types/blob/component-rebase/design/proposals/module-linking/Explainer.md#binary-format-considerations) needs more explanation as to why module linking proposal (and by extension the component model proposal) is departing from the core module spec entirely?

Reply:
We now have just `Section` (for modules) and `ComponentSection` (for components).

I've removed having to explicitly specify the format for each of the section encoders; the format is now determined by how the API is used.

Some more work is needed once the binary grammar of the spec has been updated, but the new commit at least makes the API more reasonable.